### PR TITLE
Add save controller

### PR DESCRIPTION
**Pull request by foxhatleo**

### CODE REVIEW

File: source/controllers/PPSaveController.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        LevelMetadata() : locked(true), score(0) {}
+        LevelMetadata(bool locked, uint score) {
+            this->score = score;
+            this->locked = locked;
+        }
+    };
+
+    unordered_map<string, LevelMetadata> _levels;
+
+    bool _colorblind;
```
Comment (BryanGraeser) [inline source/controllers/PPSaveController.h:24]: Why are we storing this by level?

Comment (BryanGraeser) [reply source/controllers/PPSaveController.h:24]: lol nevermind, misread

File: source/controllers/PPSaveController.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    const string &level) const {
+    auto i = _levels.find(level);
+    if (i == _levels.end()) return LevelMetadata();
+    return i->second;
+}
+
+SaveController::LevelMetadata &SaveController::_ensureLevel(
+    const string &level) {
+    return _levels[level];
+}
```
Comment (BryanGraeser) [inline source/controllers/PPSaveController.cpp:15]: This is checking is a level exists, correct?

File: source/controllers/PPSaveController.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        lv->appendValue("locked", p.second.locked);
+        lv->appendValue("score", (long)p.second.score);
+        l->appendChild(p.first, lv);
+    }
+    v->appendChild("levels", l);
+    auto w = JsonWriter::alloc(SAVE_PATH);
+    w->writeJson(v, true);
+    w->flush();
+    w->close();
+}
```
Comment (BryanGraeser) [inline source/controllers/PPSaveController.cpp:54]: What is this function doing?

File: source/controllers/PPSaveController.cpp (hunk 2 above)
Comment (BryanGraeser) [inline source/controllers/PPSaveController.cpp:15]: Why is this called "ensureLevel?" Isn't it just "getLevelMetadata?"

**Review by BryanGraeser — APPROVED** [review]
Looks good to me. Had a couple questions about specific functions just for understanding. Sorry for the delay

File: source/controllers/PPSaveController.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+void SaveController::_load() {
+    auto r = JsonReader::alloc(SAVE_PATH);
+    if (r == nullptr) return;
+    json_t v = r->readJson();
+    if (v->getInt("version", 1) == 1) {
+        _colorblind = v->getBool("colorblind", false);
+        _sfxVolume = v->getFloat("sfxVolume", 1);
+        _bgmVolume = v->getFloat("bgmVolume", 1);
```
Comment (AnshG714) [inline source/controllers/PPSaveController.cpp:24]: can be done in the future but maybe there can be a setting to set whether the palette is on the right or on the left?

File: source/controllers/PPSaveController.cpp (hunk 3 above)
Comment (AnshG714) [reply source/controllers/PPSaveController.cpp:54]: Think this just writes the most updated values of `_colorblind`, `_sfxVolume`, `_bgmVolume` etc. to memory

File: source/controllers/PPSaveController.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
+#include "PPSaveController.h"
+
+#define SAVE_PATH (Application::get()->getSaveDirectory() + "save")
+
+SaveController::LevelMetadata SaveController::_getLevel(
```
Comment (AnshG714) [inline source/controllers/PPSaveController.cpp:5]: should this be renamed to `_getLevelMetaData`?

**Review by AnshG714 — APPROVED** [review]
Added a few minor things but lgtm

**Pull request merged**